File: charges.js
// charges.js — Indian index-options round-trip cost model.
//
// Single source of truth for transaction costs so the pre-trade charge-aware
// guard and any P&L accounting use the SAME math (no divergence). Rates are the
// standard NSE/BSE equity-options schedule (premium-based), tunable via .env.
//
// A "round trip" = one BUY leg + one SELL leg on a long-option position.
//
// All rates are applied on the PREMIUM TURNOVER (price × quantity), which is how
// option charges actually work in India — NOT on notional/strike value.

// ── Default rates (override via .env) ──────────────────────────────
// Brokerage: flat per executed order (₹20 typical discount broker), charged
//   on BOTH legs → ₹40 round trip by default.
const BROKERAGE_PER_ORDER = parseFloat(process.env.CHARGE_BROKERAGE_PER_ORDER || 20);
// STT: 0.1% on the SELL-side premium only (options, as of current schedule).
const STT_SELL_PCT        = parseFloat(process.env.CHARGE_STT_SELL_PCT        || 0.1) / 100;
// Exchange transaction charge: ~0.03503% (NSE options) on premium, both sides.
const EXCH_TXN_PCT        = parseFloat(process.env.CHARGE_EXCH_TXN_PCT        || 0.03503) / 100;
// SEBI turnover fee: ₹10 per crore = 0.0001% on premium, both sides.
const SEBI_PCT            = parseFloat(process.env.CHARGE_SEBI_PCT            || 0.0001) / 100;
// Stamp duty: 0.003% on the BUY-side premium only.
const STAMP_BUY_PCT       = parseFloat(process.env.CHARGE_STAMP_BUY_PCT       || 0.003) / 100;
// GST: 18% on (brokerage + exchange txn + SEBI).
const GST_PCT             = parseFloat(process.env.CHARGE_GST_PCT             || 18) / 100;

/**
 * Round-trip transaction cost for a long-option position.
 * @param {number} entryPrice  premium paid per unit (buy)
 * @param {number} exitPrice   premium received per unit (sell). Defaults to entryPrice
 *                             when unknown (pre-trade estimate uses entry as proxy).
 * @param {number} quantity    total units (lots × lotSize)
 * @returns {{ total:number, breakdown:object }}
 */
function roundTripCharges(entryPrice, exitPrice, quantity) {
  const buyTurnover  = Math.max(0, entryPrice) * quantity;
  const sellTurnover = Math.max(0, (exitPrice ?? entryPrice)) * quantity;

  const brokerage = BROKERAGE_PER_ORDER * 2;                 // both legs
  const stt       = sellTurnover * STT_SELL_PCT;             // sell only
  const exchTxn   = (buyTurnover + sellTurnover) * EXCH_TXN_PCT;
  const sebi      = (buyTurnover + sellTurnover) * SEBI_PCT;
  const stamp     = buyTurnover * STAMP_BUY_PCT;             // buy only
  const gst       = (brokerage + exchTxn + sebi) * GST_PCT;

  const total = brokerage + stt + exchTxn + sebi + stamp + gst;
  return {
    total: +total.toFixed(2),
    breakdown: {
      brokerage: +brokerage.toFixed(2),
      stt:       +stt.toFixed(2),
      exchTxn:   +exchTxn.toFixed(2),
      sebi:      +sebi.toFixed(4),
      stamp:     +stamp.toFixed(2),
      gst:       +gst.toFixed(2)
    }
  };
}

module.exports = { roundTripCharges };

File: antigravity-py/app/charges.py
"""
Indian index-options F&O charges. ALWAYS net these (plus slippage) out of P&L —
never hide them. Rates as of FY2024-25 for NSE/BSE options.

A "leg" here = one option order (buy or sell) of `qty` units at `price`.
"""
from __future__ import annotations

BROKERAGE_PER_ORDER = 20.0      # flat ₹20/order (discount-broker standard); ₹0 for many
STT_SELL_PREMIUM = 0.000625     # 0.0625% on SELL-side premium (options)
EXCHANGE_TXN = 0.00035          # ~0.035% of premium (NSE options; BSE similar)
SEBI_FEES = 0.000001            # ₹10 per crore
GST = 0.18                      # 18% on (brokerage + exchange txn + sebi)
STAMP_BUY = 0.00003             # 0.003% on BUY-side premium


def leg_charges(price: float, qty: int, side: str) -> float:
    """Total statutory + brokerage cost for a single option leg."""
    turnover = price * qty
    brokerage = BROKERAGE_PER_ORDER
    exch = EXCHANGE_TXN * turnover
    sebi = SEBI_FEES * turnover
    gst = GST * (brokerage + exch + sebi)
    stt = STT_SELL_PREMIUM * turnover if side.upper() == "SELL" else 0.0
    stamp = STAMP_BUY * turnover if side.upper() == "BUY" else 0.0
    return round(brokerage + exch + sebi + gst + stt + stamp, 2)


def round_trip(entry_price: float, exit_price: float, qty: int, position: str) -> float:
    """
    Charges for a full round trip.
      position 'SELL' = short (sell to open, buy to close) — the strangle/condor case.
      position 'BUY'  = long  (buy to open, sell to close) — directional buying.
    """
    if position.upper() == "SELL":
        return leg_charges(entry_price, qty, "SELL") + leg_charges(exit_price, qty, "BUY")
    return leg_charges(entry_price, qty, "BUY") + leg_charges(exit_price, qty, "SELL")


def slippage(price: float, qty: int, pct: float = 0.03) -> float:
    """Per-fill slippage estimate (default 3% per fill — the backtest stress level)."""
    return round(price * qty * pct, 2)
